core/common/src/main/java/alluxio/util/executor/UniqueBlockingQueue.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.util.executor;
+
+import alluxio.collections.ConcurrentHashSet;
+
+import java.util.AbstractQueue;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A blocking queue containing only unique elements, based on LinkedBlockingQueue implementation.
+ *
+ * We serialize the insertion into the queue, otherwise, we may end up with duplicate elements
+ * in the queue.
+ *
+ * @param <T> element type
+ */
+public class UniqueBlockingQueue<T> extends AbstractQueue<T> implements BlockingQueue<T> {
+  private ConcurrentHashSet<T> mElementSet = new ConcurrentHashSet<>();
+  private BlockingQueue<T> mBlockingQueue;
+
+  /**
+   * Constructor for a UniqueBlockingQueue.
+   *
+   * @param capacity capacity of the blocking queue
+   */
+  public UniqueBlockingQueue(int capacity) {
+    mBlockingQueue = new LinkedBlockingQueue<>(capacity);
```

Comment:
Should we initialize `mElementSet` here with the capacity as well? Although the concept of capacity (fixed vs initial) is different, I think the set is bounded by capacity as well.

Reply:
Fair point, let's keep it as is then.